src/main/java/de/tum/cit/aet/artemis/core/service/LLMTokenUsageService.java

```diff
@@ -0,0 +1,70 @@
+package de.tum.cit.aet.artemis.core.service;
+
+import static de.tum.cit.aet.artemis.core.config.Constants.PROFILE_IRIS;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.springframework.context.annotation.Profile;
+import org.springframework.stereotype.Service;
+
+import de.tum.cit.aet.artemis.core.domain.Course;
+import de.tum.cit.aet.artemis.core.domain.LLMTokenUsage;
+import de.tum.cit.aet.artemis.core.domain.User;
+import de.tum.cit.aet.artemis.core.repository.LLMTokenUsageRepository;
+import de.tum.cit.aet.artemis.exercise.domain.Exercise;
+import de.tum.cit.aet.artemis.iris.domain.message.IrisMessage;
+import de.tum.cit.aet.artemis.iris.service.pyris.dto.data.PyrisLLMCostDTO;
+import de.tum.cit.aet.artemis.iris.service.pyris.job.PyrisJob;
+
+/**
+ * Service for managing the LLMTokenUsage by all LLMs in Artemis
+ */
+@Service
+@Profile(PROFILE_IRIS)
+public class LLMTokenUsageService {
+
+    private final LLMTokenUsageRepository llmTokenUsageRepository;
+
+    public LLMTokenUsageService(LLMTokenUsageRepository llmTokenUsageRepository) {
+        this.llmTokenUsageRepository = llmTokenUsageRepository;
+    }
+
+    /**
+     * method saves the token usage to the database with a link to the IrisMessage
+     * messages of the same job are grouped together by saving the job id as a trace id
+     *
+     * @param job      used to create a unique traceId to group multiple LLM calls
+     * @param message  IrisMessage to map the usage to an IrisMessage
+     * @param exercise to map the token cost to an exercise
+     * @param user     to map the token cost to a user
+     * @param course   to map the token to a course
+     * @param tokens   token cost lsit of type PyrisLLMCostDTO
+     * @return list of the saved data
+     */
+    public List<LLMTokenUsage> saveIrisTokenUsage(PyrisJob job, IrisMessage message, Exercise exercise, User user, Course course, List<PyrisLLMCostDTO> tokens) {
```

Comment:
The method has multiple fields which can be null. I think we can use a builder pattern here to make the code more readable and extendable. 

```suggestion
    public List<LLMTokenUsage> saveIrisTokenUsage(Function<IrisTokenUsageBuilder, IrisTokenUsageBuilder> builderFunction) {
        IrisTokenUsageBuilder builder = builderFunction.apply(new IrisTokenUsageBuilder());
        // Implementation of saving token usage
        // Use builder.getJob(), builder.getMessage(), etc. to access the values
        // ...
    }
```

You can define a static builder class inside the service:
```[java]
public static class IrisTokenUsageBuilder {
        private PyrisJob job;
        private Course course;
        private List<PyrisLLMCostDTO> tokens;
        private Optional<IrisMessage> message = Optional.empty();
        private Optional<Exercise> exercise = Optional.empty();
        private Optional<User> user = Optional.empty();

        public IrisTokenUsageBuilder withJob(PyrisJob job) {
            this.job = job;
            return this;
        }

        public IrisTokenUsageBuilder withCourse(Course course) {
            this.course = course;
            return this;
        }

        public IrisTokenUsageBuilder withTokens(List<PyrisLLMCostDTO> tokens) {
            this.tokens = tokens;
            return this;
        }

        public IrisTokenUsageBuilder withMessage(IrisMessage message) {
            this.message = Optional.ofNullable(message);
            return this;
        }

        public IrisTokenUsageBuilder withExercise(Exercise exercise) {
            this.exercise = Optional.ofNullable(exercise);
            return this;
        }

        public IrisTokenUsageBuilder withUser(User user) {
            this.user = Optional.ofNullable(user);
            return this;
        }

        // Getters
        public PyrisJob getJob() { return job; }
        public Course getCourse() { return course; }
        public List<PyrisLLMCostDTO> getTokens() { return tokens; }
        public Optional<IrisMessage> getMessage() { return message; }
        public Optional<Exercise> getExercise() { return exercise; }
        public Optional<User> getUser() { return user; }
    }
```

You can then use the method like in the following:
```[java]
 List<LLMTokenUsage> result = service.saveIrisTokenUsage(builder -> 
    builder.withJob(job)
           .withCourse(course)
           .withTokens(tokens)
           .withUser(user)
           .withMessage(message)
);
```

Reply:
Thank you for the feedback. The builder pattern has been implemented, and the code has been updated!